lib/datadog/tracing.rb

```diff
@@ -0,0 +1,198 @@
+module Datadog
+  # Datadog APM tracing public API.
+  #
+  # The Datadog teams ensures that public methods in this module
+  # only receive backwards compatible changes, and breaking changes
+  # will only occur in new major versions releases.
+  # @public_api
+  module Tracing
+    class << self
+      # (see Datadog::Tracer#trace)
+      # @public_api
+      def trace(name, continue_from: nil, **span_options, &block)
+        tracer.trace(name, continue_from: continue_from, **span_options, &block)
+      end
+
+      # (see Datadog::Tracer#continue_trace!)
+      # @public_api
+      def continue_trace!(digest, &block)
+        tracer.continue_trace!(digest, &block)
+      end
+
+      # The currently active {Datadog::Tracer} instance.
+      #
+      # The instance returned can change throughout the lifetime of the application.
+      # This means it is not advisable to cache it.
+      #
+      # The trace can be configured through {.configure},
+      # through {Datadog::Configuration::Settings::DSL::Tracer} options.
+      #
+      # TODO: This next paragraph can be better written.
+      #
+      # Most of the functionality available through the {.tracer} instance is
+      # also available in public methods in the {Datadog::Tracing} module.
+      # It is preferable to use the public methods in the {Datadog::Tracing} when possible
+      # as {Datadog::Tracing} strongly defines the tracing public API, and thus
+      # we strive to no introduce breaking changes to {Datadog::Tracing} methods.
+      #
+      # @return [Datadog::Tracer] the active tracer
+      # @!attribute [r] tracer
+      # @public_api
+      def tracer
+        components.tracer
+      end
+
+      # The tracer's internal logger instance.
+      # All tracing log output is handled by this object.
+      #
+      # The logger can be configured through {.configure},
+      # through {Datadog::Configuration::Settings::DSL::Logger} options.
+      #
+      # @!attribute [r] logger
+      # @public_api
+      def logger
+        Datadog.logger
+      end
+
+      # Current tracer configuration.
+      #
+      # To modify the configuration, use {.configure}.
+      #
+      # @return [Datadog::Configuration::Settings]
+      # @!attribute [r] configuration
+      # @public_api
+      def configuration
+        Datadog.configuration
+      end
```

Comment:
I've discussed this with David some time ago, and because we expose the whole configuration API during `Datadog.configure {|c|}` calls (including options/settings reader methods), it would make sense to have that same object available for user introspection.

This is actually used today by integrations, like in https://github.com/DataDog/dd-trace-rb/blob/668146c9e8c497049b3efc56f02eae66b76e2e2a/lib/ddtrace/contrib/rake/instrumentation.rb#L85. It's the main API used to access integration configuration, if one is writing their own integration.

Reply:
Good points. I guess we could change this, but probably not for 1.x.